tests/unit/commands/local/lib/test_api_collector.py

```diff
@@ -0,0 +1,131 @@
+from unittest import TestCase
+from parameterized import parameterized
+
+from samcli.lib.providers.api_collector import ApiCollector
+from samcli.local.apigw.local_apigw_service import Route, Authorizer
+
+
+class TestApiCollector_linking_authorizer(TestCase):
+    def setUp(self):
+        self.apigw_id = "apigw1"
+
+        self.api_collector = ApiCollector()
+
+    @parameterized.expand(
+        [
+            (  # test link default authorizer
+                [Route(function_name="func1", path="path1", methods=["get"], stack_path="path1", authorizer_name="")],
+                {
+                    "auth1": Authorizer(authorizer_name="auth1", type="token1", payload_version="1.0"),
+                    "auth2": Authorizer(authorizer_name="auth2", type="token2", payload_version="1.0"),
+                },
+                "auth1",
+                [
+                    Route(
+                        function_name="func1",
+                        path="path1",
+                        methods=["get"],
+                        stack_path="path1",
+                        authorizer_name="auth1",
+                        authorizer_object=Authorizer(authorizer_name="auth1", type="token1", payload_version="1.0"),
+                    )
+                ],
+            ),
+            (  # test link non existant default authorizer
+                [Route(function_name="func1", path="path1", methods=["get"], stack_path="path1", authorizer_name="")],
+                {
+                    "auth1": Authorizer(authorizer_name="auth1", type="token1", payload_version="1.0"),
+                    "auth2": Authorizer(authorizer_name="auth2", type="token2", payload_version="1.0"),
+                },
+                None,
+                [
+                    Route(
+                        function_name="func1",
+                        path="path1",
+                        methods=["get"],
+                        stack_path="path1",
+                        authorizer_name=None,
+                        authorizer_object=None,
+                    )
+                ],
+            ),
+            (  # test no authorizer defined in route
+                [Route(function_name="func1", path="path1", methods=["get"], stack_path="path1", authorizer_name=None)],
+                {
+                    "auth1": Authorizer(authorizer_name="auth1", type="token1", payload_version="1.0"),
+                    "auth2": Authorizer(authorizer_name="auth2", type="token2", payload_version="1.0"),
+                },
+                "auth1",
+                [
+                    Route(
+                        function_name="func1",
+                        path="path1",
+                        methods=["get"],
+                        stack_path="path1",
+                        authorizer_name=None,
```

Comment:
Discussed in above comment:
> discussed offline, this is because of how the parser is set up. The name is output in an intermediate step, and it's only later that the full object is added.